cuequivariance_torch/cuequivariance_torch/primitives/segmented_polynomial_naive.py

```diff
@@ -0,0 +1,415 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import math
+from functools import partial
+from typing import Dict, List, Optional, OrderedDict
+
+import torch
+import torch.fx
+import torch.nn as nn
+
+import cuequivariance as cue
+
+logger = logging.getLogger(__name__)
+
+
+def prod(numbers: List[int]):
+    """
+    This method is a workaround for script() not recognizing math.prod()
+    """
+    if torch.jit.is_scripting():
+        product = 1
+        for num in numbers:
+            product *= num
+        return product
+    else:
+        return math.prod(numbers)
+
+
+def to_notypeconv(t, *args, **kwargs):
+    new_kwargs = kwargs.copy()
+    new_kwargs.pop("dtype", None)
+    new_args = [None if isinstance(a, torch.dtype) else a for a in args]
+    result = t.__original_to(*new_args, **new_kwargs)
+    return result
+
+
+def disable_type_conv(t):
+    """
+    This modifier can be used on Tensors or whole Modules
+    to prevent them from being modified during to(dtype=x) calls
+    """
+    t.__original_to = t.to
+    t.to = partial(to_notypeconv, t)
+    return t
+
+
+def _tensor_product_fx(
+    descriptor: cue.SegmentedTensorProduct,
+    device: Optional[torch.device],
+    math_dtype: Optional[torch.dtype],
+    optimize_einsums: bool,
+) -> torch.nn.Module:
+    """
+    batch support of this function:
+    - at least one input operand should have a batch dimension (ndim=2)
+    - the output operand will have a batch dimension (ndim=2)
+    """
+    descriptor = descriptor.remove_zero_paths()
+    descriptor = descriptor.remove_empty_segments()
+
+    num_inputs = descriptor.num_operands - 1
+
+    if num_inputs > 0 and descriptor.num_paths > 0:
+        graph = torch.fx.Graph()
+        tracer = torch.fx.proxy.GraphAppendingTracer(graph)
+        constants = OrderedDict()
+
+        inputs = [
+            torch.fx.Proxy(graph.placeholder(f"input_{i}"), tracer)
+            for i in range(num_inputs)
+        ]
+
+        operand_subscripts = [f"Z{ss}" for ss in descriptor.subscripts.operands]
+
+        formula = (
+            ",".join([descriptor.coefficient_subscripts] + operand_subscripts[:-1])
+            + "->"
+            + operand_subscripts[-1]
+        )
+        slices = [ope.segment_slices() for ope in descriptor.operands]
+
+        outputs = []
+        for path_idx, path in enumerate(descriptor.paths):
+            segments = []
+            for oid in range(num_inputs):
+                seg_shape = descriptor.get_segment_shape(oid, path)
+                inp = inputs[oid][..., slices[oid][path.indices[oid]]]
+                if len(seg_shape) > 0:
+                    inp = inp.reshape(inputs[oid].shape[:-1] + seg_shape)
+                else:
+                    inp = inp.reshape(inputs[oid].shape[:-1])
+
+                if math_dtype is not None:
```

Comment:
Ok so I see that now `math_dtype=None` is an actual new behavior! That's cool :)
So we need to make sure that this new behavior is respected by all the method and raise a NotImplemented error otherwise.